### PR TITLE
WFLY-5550 Distributable Web subsystem

### PR DESCRIPTION
**Pull request by pferraro**
https://issues.jboss.org/browse/WFLY-5550
https://issues.jboss.org/browse/WFLY-6944
https://issues.jboss.org/browse/WFLY-5497

Requires: https://github.com/wildfly-clustering/wildfly-core/pull/1

### CODE REVIEW

File: clustering/web/container/pom.xml [hunk 1]
```diff
@@ -0,0 +1,50 @@
+<?xml version="1.0" encoding="UTF-8"?>
+<!--
+  ~ JBoss, Home of Professional Open Source.
+  ~ Copyright 2010, Red Hat, Inc., and individual contributors
```
Comment (rhusar) [inline clustering/web/container/pom.xml]: 2018

File: clustering/web/container/src/main/java/org/wildfly/clustering/web/container/HostSingleSignOnManagementProvider.java [hunk 2]
```diff
@@ -20,31 +20,23 @@
...
-public interface DistributableHostSingleSignOnManagerServiceConfiguratorProvider {
-
-    Optional<DistributableHostSingleSignOnManagerServiceConfiguratorProvider> INSTANCE = StreamSupport.stream(ServiceLoader.load(DistributableHostSingleSignOnManagerServiceConfiguratorProvider.class, DistributableHostSingleSignOnManagerServiceConfiguratorProvider.class.getClassLoader()).spliterator(), false).findFirst();
+public interface HostSingleSignOnManagementProvider {
 
     /**
      * Returns a builder of a SingleSignOnManagerFactory service for a host.
      * @param name the service name
      * @param serverName the name of the target server
      * @param hostName the name of the target host
```
Comment (rhusar) [inline clustering/web/container/src/main/java/org/wildfly/clustering/web/container/HostSingleSignOnManagementProvider.java]: javadoc is now out of date

File: clustering/web/extension/pom.xml [hunk 3]
```diff
@@ -0,0 +1,83 @@
+<?xml version="1.0" encoding="UTF-8"?>
+<!--
+  ~ JBoss, Home of Professional Open Source.
+  ~ Copyright 2010, Red Hat, Inc., and individual contributors
```
Comment (rhusar) [inline clustering/web/extension/pom.xml]: 2018

File: clustering/web/extension/pom.xml [hunk 4]
```diff
@@ -0,0 +1,83 @@
...
+
+        <dependency>
+            <groupId>org.wildfly</groupId>
+            <artifactId>wildfly-clustering-common</artifactId>
+            <version>${project.version}</version>
+            <scope>test</scope>
+            <classifier>tests</classifier>
+        </dependency>
+        <dependency>
+            <groupId>org.wildfly</groupId>
```
Comment (rhusar) [inline clustering/web/extension/pom.xml]: ${project.groupId}

File: clustering/web/extension/pom.xml [hunk 5]
```diff
@@ -0,0 +1,83 @@
...
+        </dependency>
+
+        <dependency>
+            <groupId>org.kohsuke.metainf-services</groupId>
+            <artifactId>metainf-services</artifactId>
+            <scope>provided</scope>
+        </dependency>
+
+        <dependency>
+            <groupId>org.wildfly</groupId>
```
Comment (rhusar) [inline clustering/web/extension/pom.xml]: ${project.groupId}

File: clustering/web/extension/src/main/java/org/wildfly/extension/clustering/web/SSOManagementServiceConfigurator.java [hunk 6]
```diff
@@ -0,0 +1,52 @@
...
+import org.jboss.as.clustering.controller.CapabilityServiceNameProvider;
+import org.jboss.as.clustering.controller.ResourceServiceConfigurator;
+import org.jboss.as.controller.PathAddress;
+import org.jboss.msc.Service;
+import org.jboss.msc.service.ServiceBuilder;
+import org.jboss.msc.service.ServiceTarget;
+import org.wildfly.clustering.web.sso.DistributableSSOManagementProvider;
+
+/**
+ * Abstract service configurator for single sign-on mangement providers.
```
Comment (rhusar) [inline clustering/web/extension/src/main/java/org/wildfly/extension/clustering/web/SSOManagementServiceConfigurator.java]: management*

File: clustering/web/extension/src/main/java/org/wildfly/extension/clustering/web/DistributableWebResourceDefinition.java [hunk 7]
```diff
@@ -0,0 +1,139 @@
...
+    }
+
+    @Override
+    public void accept(DeploymentProcessorTarget target) {
+        JBossAllXmlParserRegisteringProcessor.Builder builder = JBossAllXmlParserRegisteringProcessor.builder();
+        for (DistributableWebDeploymentSchema schema : EnumSet.allOf(DistributableWebDeploymentSchema.class)) {
+            builder.addParser(schema.getRoot(), DistributableWebDeploymentDependencyProcessor.CONFIGURATION_KEY, new DistributableWebDeploymentXMLReader(schema));
+        }
+        target.addDeploymentProcessor(DistributableWebExtension.SUBSYSTEM_NAME, Phase.STRUCTURE, Phase.STRUCTURE_REGISTER_JBOSS_ALL_DISTRIBUTABLE_WEB, builder.build());
+        target.addDeploymentProcessor(DistributableWebExtension.SUBSYSTEM_NAME, Phase.PARSE, Phase.PARSE_DISTRIBUTABLE_WEB, new DistributableWebDeploymentParsingProcessor());
```
[the remark this answers is not in the document]
Comment (pferraro) [reply clustering/web/extension/src/main/java/org/wildfly/extension/clustering/web/DistributableWebResourceDefinition.java]: Yes - there is.  I'll create a wildfly-clustering fork for wildfly-core.

File: clustering/web/extension/src/main/java/org/wildfly/extension/clustering/web/DistributableWebResourceDefinition.java [hunk 8]
```diff
@@ -0,0 +1,139 @@
...
+        private final AttributeDefinition definition;
+
+        Attribute(String name, String group, ModelType type, CapabilityReferenceRecorder reference) {
+            this.definition = new SimpleAttributeDefinitionBuilder(name, type)
+                    .setAttributeGroup(group)
+                    .setAllowExpression(false)
+                    .setRequired(true)
+                    .setCapabilityReference(reference)
+                    .setFlags(Flag.RESTART_RESOURCE_SERVICES)
+                    .setXmlName("default")
```
[the remark this answers is not in the document]
Comment (pferraro) [reply clustering/web/extension/src/main/java/org/wildfly/extension/clustering/web/DistributableWebResourceDefinition.java]: Yes. This is no different than default-stack vs default-channel in the jgroups subsystem.
The parser uses elementPerGroup or whatever its called.

File: clustering/web/extension/src/main/java/org/wildfly/extension/clustering/web/SSOManagementServiceConfigurator.java (hunk 6 above)
Comment (pferraro) [reply clustering/web/extension/src/main/java/org/wildfly/extension/clustering/web/SSOManagementServiceConfigurator.java]: :) Typing the same word 5000 times one is bound to make a typo...

File: clustering/web/extension/src/main/java/org/wildfly/extension/clustering/web/DistributableWebResourceDefinition.java (hunk 7 above)
[the remark this answers is not in the document]
Comment (pferraro) [reply clustering/web/extension/src/main/java/org/wildfly/extension/clustering/web/DistributableWebResourceDefinition.java]: https://github.com/wildfly-clustering/wildfly-core/pull/1

File: clustering/web/extension/src/main/java/org/wildfly/extension/clustering/web/deployment/DistributableWebDeploymentDependencyProcessor.java [hunk 9]
```diff
@@ -0,0 +1,95 @@
...
+import org.jboss.msc.Service;
+import org.jboss.msc.service.ServiceBuilder;
+import org.jboss.msc.service.ServiceController;
+import org.jboss.msc.service.ServiceName;
+import org.jboss.msc.service.ServiceTarget;
+import org.wildfly.clustering.web.WebProviderRequirement;
+import org.wildfly.clustering.web.session.DistributableSessionManagementProvider;
+
+/**
+ * {@link DeploymentUnitProcessor} that establishes the requisite {link DistributableSessionManagementProvider} dependency
```
Comment (rhusar) [inline clustering/web/extension/src/main/java/org/wildfly/extension/clustering/web/deployment/DistributableWebDeploymentDependencyProcessor.java]: @link

File: clustering/web/undertow/src/main/java/org/wildfly/clustering/web/undertow/session/UndertowSessionManagementProviderFactory.java [hunk 10]
```diff
@@ -0,0 +1,70 @@
...
+        this.legacyFactory = factories.next();
+    }
+
+    @Override
+    public SessionManagementProvider createSessionManagementProvider(DeploymentUnit unit, ReplicationConfig config) {
+        DistributableSessionManagementProvider provider = unit.getAttachment(DistributableSessionManagementProvider.ATTACHMENT_KEY);
+        // For compatibility, honor legacy <replication-config/> over an attached provider
+        if ((config != null) || (provider == null)) {
+            if (provider != null) {
+                UndertowClusteringLogger.ROOT_LOGGER.debugf("Legacy <replication-config/> overriding attached distributable session management provider for %s", unit.getName());
```
[the remark this answers is not in the document]
Comment (pferraro) [reply clustering/web/undertow/src/main/java/org/wildfly/clustering/web/undertow/session/UndertowSessionManagementProviderFactory.java]: Good point.

File: feature-pack/src/main/resources/modules/system/layers/base/org/wildfly/extension/clustering/web/main/module.xml [hunk 11]
```diff
@@ -0,0 +1,65 @@
...
+  ~ MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU
+  ~ Lesser General Public License for more details.
+  ~
+  ~ You should have received a copy of the GNU Lesser General Public
+  ~ License along with this software; if not, write to the Free
+  ~ Software Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA
+  ~ 02110-1301 USA, or see the FSF site: http://www.fsf.org.
+  -->
+
+<module xmlns="urn:jboss:module:1.5" name="org.wildfly.extension.clustering.web">
```
[the remark this answers is not in the document]
Comment (pferraro) [reply feature-pack/src/main/resources/modules/system/layers/base/org/wildfly/extension/clustering/web/main/module.xml]: Are we at 1.8 already?

[the remark this answers is not in the document]
Comment (rhusar) [reply feature-pack/src/main/resources/modules/system/layers/base/org/wildfly/extension/clustering/web/main/module.xml]: Yeah https://github.com/jboss-modules/jboss-modules/blob/1.x/src/main/resources/schema/module-1_8.xsd

File: docs/src/main/asciidoc/_high-availability/subsystem-support/Distributable_Web_Applications.adoc [hunk 12]
```diff
@@ -0,0 +1,178 @@
...
+Stores all session attributes within a single cache entry.
+This is generally more expensive than ATTRIBUTE granularity, but preserves any cross-attribute object references.
+ATTRIBUTE:::
+Stores each session attribute within a separate cache entry.
+This is generally more efficient than SESSION granularity, but does not preserve any cross-attribute object references.
+routing::
+This defines the algorithm for generating the route to be appended onto the session ID (within the JSESSIONID cookie, or when encoding URLs).
+This annotation of the session ID is used by load balancers to advise how future requests for existing sessions should be directed.
+Routing is designed to be opaque to application code such that calls to HttpSession.getId() always return an unmodified session ID.
+This is only generated when creating the JSESSIONID cookie, or when encoding URLs via HttpServletResponse.encodeURL() and encodeRedirectURL().
```
Comment (rhusar) [inline docs/src/main/asciidoc/_high-availability/subsystem-support/Distributable_Web_Applications.adoc]: I prefer these formatted with monospaced fonts just by `HttpServletResponse.encodeURL()` (i.e. surround with `  )

Comment (pferraro) [reply docs/src/main/asciidoc/_high-availability/subsystem-support/Distributable_Web_Applications.adoc]: good point

**Comment by rhusar** [comment]
@pferraro did you really mean to merge this into master? :-)

**Comment by pferraro** [comment]
I intended to merge it into a topic branch.  Oops.

**Comment by rhusar** [comment]
@pferraro are you planning to then open a PR from this repo to upstream? I think that's another habit that could be of interest especially during vacation times.

**Pull request merged**